engine/access/rpc/backend/backend.go

```diff
@@ -226,12 +230,80 @@ func (b *Backend) GetNetworkParameters(_ context.Context) access.NetworkParamete
 
 // GetLatestProtocolStateSnapshot returns the latest finalized snapshot
 func (b *Backend) GetLatestProtocolStateSnapshot(_ context.Context) ([]byte, error) {
-	data, err := convert.SnapshotToBytes(b.state.Final())
+	snapshot := b.state.Final()
+
+	validSnapshot, err := b.getValidSnapshot(snapshot, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	return convert.SnapshotToBytes(validSnapshot)
+}
+
+// getValidSnapshot will return a valid snapshot that has a sealing segment which
+// 1. does not contain any blocks that span an epoch transition
+// 2. does not contain any blocks that span an epoch phase transition
+// If a snapshot does contain an invalid sealing segment query the state
+// by height of each block in the segment and return a snapshot at the point
+// where the transition happens.
+func (b *Backend) getValidSnapshot(snapshot protocol.Snapshot, blocksVisited int) (protocol.Snapshot, error) {
+	if blocksVisited > snapshotHistoryLimit {
+		return nil, fmt.Errorf("failed to get valid snapshot reached look back limit ")
+	}
+
+	segment, err := snapshot.SealingSegment()
+	if err != nil {
+		return nil, err
+	}
+
+	// get snapshots at the highest and lowest blocks in segment for comparison
+	snapshotAtHighest := b.state.AtHeight(segment.Highest().Header.Height)
+	snapshotAtLowest := b.state.AtHeight(segment.Lowest().Header.Height)
+
+	counterAtHighest, err := snapshotAtHighest.Epochs().Current().Counter()
+	if err != nil {
+		return nil, err
+	}
+
+	phaseAtHighest, err := snapshotAtHighest.Phase()
 	if err != nil {
 		return nil, err
 	}
 
-	return data, nil
+	counterAtLowest, err := snapshotAtLowest.Epochs().Current().Counter()
+	if err != nil {
+		return nil, err
+	}
+
+	phaseAtLowest, err := snapshotAtLowest.Phase()
+	if err != nil {
+		return nil, err
+	}
+
+	// check if any epoch or phase transition occurs, visit each node
+	// in strict order of decreasing height starting at head
+	if counterAtHighest != counterAtLowest || phaseAtHighest != phaseAtLowest {
+		for i := len(segment.Blocks) - 1; i >= 0; i-- {
+			blocksVisited++
```

Comment:
better not to update the input value, instead, do:

```
return b.getValidSnapshot(snapshotAtBlock, blocksVisited + 1)
```

Reply:
The `blocksVisited` parameter variable is a primitive so I agree it is fine to modify it directly here, rather than creating a new variable.